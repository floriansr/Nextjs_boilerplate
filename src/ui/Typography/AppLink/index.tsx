import * as React from 'react';

import Link, { LinkProps } from 'next/link';

import styled, { css, CSSObject } from 'styled-components';

import AppBox, { AppBoxProps, FontWeight, LineHeight } from 'ui/AppBox';

interface OwnProps {
  readonly applyActiveLinkStyles?: boolean;
  readonly applyFilterOnHover?: boolean;
  readonly activeLinkStyles?: TemplateStringsArray | CSSObject;
}

const AppLinkInner = styled(AppBox)<OwnProps>`
  transition: filter 0.2s ease, opacity 0.2s ease, color 0.3s ease;
  &:hover {
    cursor: pointer;
    ${(props) =>
      props.applyFilterOnHover &&
      css`
        filter: brightness(80%);
      `};
  }
  ${(props) =>
    props.applyActiveLinkStyles &&
    css`
      color: ${(props) => props.theme.colors.link};
    `}
`;

type Props = OwnProps & Omit<AppBoxProps, 'as'> & LinkProps;

const AppLink = React.forwardRef<HTMLElement, Props>(
  (
    {
      href,
      as,
      replace,
      scroll,
      shallow,
      prefetch,
      activeLinkStyles,
      ...rest
    }: Props,
    ref
  ) => {
    return (
      <Link
        href={href}
        as={as}
        replace={replace}
        scroll={scroll}
        shallow={shallow}
        passHref={true}
        prefetch={prefetch}>
        <AppLinkInner
          applyActiveLinkStyles={true}
          activeLinkStyles={activeLinkStyles}
          applyFilterOnHover={true}
          css={{ textDecoration: 'none' }}
          fontWeight={FontWeight.Semibold}
          ref={ref}
          fontFamily="body"
          color="link"
          lineHeight={LineHeight.Base}
          {...rest}
          as="a"
        />
      </Link>
    );
  }
);

AppLink.displayName = 'AppLink';

export default AppLink;
