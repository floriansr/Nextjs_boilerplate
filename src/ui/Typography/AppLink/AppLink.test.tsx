import * as React from 'react';

import { render } from '@testing-library/react';

import AppLink from './index';

describe('<AppLink />', () => {
  const TEXT = 'Hello world';

  it('AppLink tree', () => {
    const tree = render(<AppLink href="/" />);
    expect(tree).toMatchSnapshot();
  });

  it('should render as a link', () => {
    const { container } = render(<AppLink href="/">{TEXT}</AppLink>);
    expect(container.firstChild?.nodeName).toBe('A');
  });

  it('should render as A element then rerender component', () => {
    const tree = render(<AppLink href="/">{TEXT}</AppLink>);
    expect(tree.getByText(TEXT).nodeName).toBe('A');

    tree.rerender(<AppLink href="/">{TEXT}</AppLink>);
    expect(tree.getByText(TEXT).nodeName).toBe('A');
  });

  it('should render as the specified node in the as & color theme provider', () => {
    const { container } = render(<AppLink href="/" color="red" />);
    expect(container.firstChild?.nodeName).toBe('A');
    expect(container.firstChild).toHaveStyle({
      color: 'red',
      'font-weight': 'semibold',
      'font-family': 'body',
      'line-height': 'base'
    });
  });
});
