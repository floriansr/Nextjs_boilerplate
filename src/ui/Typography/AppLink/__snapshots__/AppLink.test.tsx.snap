// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppLink /> AppLink passHref 1`] = `
.c0 {
  font-weight: 600;
  font-family: Zilla Slab,-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: #2563eb;
  line-height: 1.5;
}

.c1 {
  -webkit-transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  color: #2563eb;
}

.c1:hover {
  cursor: pointer;
  -webkit-filter: brightness(80%);
  filter: brightness(80%);
}

<div>
  <a
    class="c0 c1"
    href="/"
  />
</div>
`;

exports[`<AppLink /> AppLink tree 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  font-weight: 600;
  font-family: Zilla Slab,-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: #2563eb;
  line-height: 1.5;
}

.c1 {
  -webkit-transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  color: #2563eb;
}

.c1:hover {
  cursor: pointer;
  -webkit-filter: brightness(80%);
  filter: brightness(80%);
}

<body>
    <div>
      <a
        class="c0 c1"
        href="/"
      />
    </div>
  </body>,
  "container": <div>
    <a
      class="sc-bdfBwQ sc-gsTCUz kXyOeL iWpNvF"
      href="/"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<AppLink /> color theme dark 1`] = `
.c0 {
  font-weight: 600;
  font-family: Zilla Slab,-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: #60a5fa;
  line-height: 1.5;
}

.c1 {
  -webkit-transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  color: #60a5fa;
}

.c1:hover {
  cursor: pointer;
  -webkit-filter: brightness(80%);
  filter: brightness(80%);
}

<div>
  <a
    class="c0 c1"
    href="/"
  />
</div>
`;

exports[`<AppLink /> color theme light 1`] = `
.c0 {
  font-weight: 600;
  font-family: Zilla Slab,-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: #2563eb;
  line-height: 1.5;
}

.c1 {
  -webkit-transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  transition: filter 0.2s ease,opacity 0.2s ease,color 0.3s ease;
  color: #2563eb;
}

.c1:hover {
  cursor: pointer;
  -webkit-filter: brightness(80%);
  filter: brightness(80%);
}

<div>
  <a
    class="c0 c1"
    href="/"
  />
</div>
`;
